test/unit/extensions/extensions.spec.ts

```diff
@@ -0,0 +1,174 @@
+        process.env['EXT_INSTANCE_ID'] = 'test-instance';
+        expect(() => extensions.runtime())
+          .to.throw('Expected PROJECT_ID not to be undefined in Extensions runtime enviornment');
+      });
+  
+  
+      it('should error if called without EXT_INSTANCE_ID', () => {
+        process.env['PROJECT_ID'] = 'test-project';
+        expect(() => extensions.runtime())
+          .to.throw('Runtime can only available from within a running Extension instance.');
+      });
+  
+      it('should not error if called from an extension', () => {
+        process.env['PROJECT_ID'] = 'test-project';
+        process.env['EXT_INSTANCE_ID'] = 'test-instance';
+        expect(() => extensions.runtime()).not.to.throw();
+      });
+    });
+    
+    describe('setProcessingState', () => {
+      let httpClientStub: sinon.SinonStub;
+      beforeEach(() => {
+        process.env['PROJECT_ID'] = 'test-project';
+        process.env['EXT_INSTANCE_ID'] = 'test-instance';
+        httpClientStub = sinon.stub(HttpClient.prototype, 'send');
+      });
+  
+      afterEach(() => {
+        httpClientStub.restore();
+      });
+  
+      for (const state of ['PROCESSING_FAILED', 'PROCESSING_WARNING','PROCESSING_COMPLETE', 'NONE']) {
+        it(`should set ${state} state`, async () => {
+          const expectedRuntimeData = {
+            processingState: {
+              state: state as SettableProcessingState,
+              detailMessage: 'done processing',
+            },
+          }
+          const expected =  sinon.match((req: HttpRequestConfig) => { 
+            const url = 'https://firebaseextensions.googleapis.com/' +
+              'v1beta/projects/test-project/instances/test-instance/runtimeData';
+            return req.method == 'PATCH' &&
+              req.url == url &&
+              JSON.stringify(req.data) == JSON.stringify(expectedRuntimeData);
+          }, 'Incorrect URL or Method');
+          httpClientStub.withArgs(expected).resolves(utils.responseFrom(expectedRuntimeData, 200));
+
+
+          await extensions.runtime().setProcessingState(state as SettableProcessingState, 'done processing');
+          expect(httpClientStub).to.have.been.calledOnce;
+        });
+      }
+  
+      it('should covert errors in FirebaseErrors', async () => {
+        httpClientStub.rejects(utils.errorFrom('Something went wrong', 404));
+        await expect(extensions.runtime().setProcessingState('PROCESSING_COMPLETE', 'a message'))
+          .to.eventually.be.rejectedWith(FirebaseExtensionsError);
+      });
+    });
+
+    describe('setFatalError', () => {
+      let httpClientStub: sinon.SinonStub;
+      beforeEach(() => {
+        process.env['PROJECT_ID'] = 'test-project';
+        process.env['EXT_INSTANCE_ID'] = 'test-instance';
+        httpClientStub = sinon.stub(HttpClient.prototype, 'send');
+      });
+  
+      afterEach(() => {
+        httpClientStub.restore();
+      });
+  
+      it('should set fatal error', async () => {
```

Comment:
I just realized that we probably should not allow a fatal_error with an empty error message. (I'll fix the BE to match)

Reply:
Good idea, i'll add a client side check too